news-bundle/src/EventListener/SitemapListener.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\NewsBundle\EventListener;
+
+use Contao\Controller;
+use Contao\CoreBundle\Event\ContaoCoreEvents;
+use Contao\CoreBundle\Event\SitemapEvent;
+use Contao\CoreBundle\Framework\ContaoFramework;
+use Contao\News;
+use Contao\NewsArchiveModel;
+use Contao\NewsModel;
+use Contao\PageModel;
+use Terminal42\ServiceAnnotationBundle\Annotation\ServiceTag;
+
+/**
+ * @ServiceTag("kernel.event_listener", event=ContaoCoreEvents::SITEMAP)
+ *
+ * @internal
+ */
+class SitemapListener
+{
+    private ContaoFramework $framework;
+
+    public function __construct(ContaoFramework $framework)
+    {
+        $this->framework = $framework;
+    }
+
+    public function __invoke(SitemapEvent $sitemapEvent): void
+    {
+        $sitemap = $sitemapEvent->getDocument();
+        $urlSet = $sitemap->firstChild;
+
+        $this->framework->initialize();
+
+        /** @var NewsArchiveModel $newsArchiveModel */
+        $newsArchiveModelAdapter = $this->framework->getAdapter(NewsArchiveModel::class);
+
+        // Get all news archives
+        $archives = $newsArchiveModelAdapter->findAll();
+
+        if (null === $archives) {
+            return;
+        }
+
+        /** @var PageModel $pageModelAdapter */
+        $pageModelAdapter = $this->framework->getAdapter(PageModel::class);
+
+        /** @var Controller $controllerAdapter */
+        $controllerAdapter = $this->framework->getAdapter(Controller::class);
+
+        /** @var NewsModel $newsModelAdapter */
+        $newsModelAdapter = $this->framework->getAdapter(NewsModel::class);
+
+        /** @var News $newsAdapter */
+        $newsAdapter = $this->framework->getAdapter(News::class);
+
+        $rootPageIds = array_map('intval', $sitemapEvent->getRootPageIds());
+
+        /** @var NewsArchiveModel $archive */
+        foreach ($archives as $archive) {
+            // Skip archives without a target page
+            if (empty($archive->jumpTo)) {
+                continue;
+            }
+
+            // Skip archive if protected and not accessible
+            if (!$controllerAdapter->isVisibleElement($archive)) {
```

Comment:
this method was not meant to be used here. We could use `ModuleNews::sortOutProtected()` instead (maybe make it `public static`) or use the security voter instead.

Reply:
> this method was not meant to be used here.

Why not? `Controller::isVisibleElement` is not restricted to any particular model. It is applied, and can be applied, to any model that contains the following fields:

* `protected`
* `groups`

Currently it is applied to instances of `ArticleModel`, `ContentModel` and `ModuleModel` and I do not see any reason not to apply it here as well, as it already covers everything that is necessary and also implements the `isVisibleElement` hook, which is still not deprecated.

We might of course decide to deprecate this method and its hook - but that is unrelated to this PR.